crates/astria-sequencer-relayer/src/relayer.rs

```diff
@@ -161,21 +344,129 @@ impl Relayer {
     ///
     /// `Relayer::run` never returns an error. The return type is
     /// only set to `eyre::Result` for convenient use in `SequencerRelayer`.
-    pub(crate) async fn run(self) -> eyre::Result<()> {
-        let mut interval = time::interval(self.sequencer_poll_period);
-        loop {
-            interval.tick().await;
-            match self.get_and_submit_latest_block().await {
-                Err(e) => warn!(error = ?e, "failed to get latest block from sequencer"),
-                Ok(new_state) if new_state != *self.state_tx.borrow() => {
-                    _ = self.state_tx.send_replace(new_state);
+    #[instrument(name = "Relayer::run", skip_all)]
+    pub(crate) async fn run(mut self) -> eyre::Result<()> {
+        let wait_for_da = self.wait_for_data_availability_layer(5, Duration::from_secs(5), 2.0);
+        let wait_for_seq = self.wait_for_sequencer(5, Duration::from_secs(5), 2.0);
+        match tokio::try_join!(wait_for_da, wait_for_seq) {
+            Ok(((), ())) => {}
+            Err(err) => return Err(err).wrap_err("failed to start relayer"),
+        }
+        // .await
+        // .wrap_err("failed establishing connection to data availability layer")?;
+        self.wait_for_sequencer(5, Duration::from_secs(5), 2.0)
+            .await
+            .wrap_err("failed establishing connection to the sequencer")?;
+
+        let mut sequencer_interval = interval(self.sequencer_poll_period);
+        // the first tick resolves immediately, so do it now because we don't
+        // want to poll right after start.
+        sequencer_interval.tick().await;
+
+        let stop_msg = loop {
+            select!(
+                // Query sequencer for the latest block if no task is in flight
+                _ = sequencer_interval.tick() => self.handle_sequencer_tick(),
+
+                // Handle the sequencer response by converting it
+                //
+                // NOTE: + wrapping the task in an async block makes this lazy;
+                //       + `unwrap`ping can't fail because this branch is disabled if `None`
+                res = async { self.sequencer_task.as_mut().unwrap().await }, if self.sequencer_task.is_some() => {
```

Comment:
what does the `async { self.sequencer_task.as_mut().unwrap().await }, if self.sequencer_task.is_some()` syntax do? does it only execute the future if `self.sequencer_task.is_some()`?

Reply:
Yes, the `, if <condition>` syntax only conditionally enables this `select!` branch. Moving the `.as_mut().unwrap()` into an ad-hoc async block ensure that it is only unwrapped lazily (otherwise this might panic).